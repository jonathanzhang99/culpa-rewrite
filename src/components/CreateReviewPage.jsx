import React, { useEffect, useReducer, useRef } from "react";
import { useHistory } from "react-router-dom";
import { Divider, List, Segment, Header, Grid } from "semantic-ui-react";

import Form, { FormGroup, SubmitConfirm } from "components/common/Form";
import {
  DropdownInput,
  RadioInputGroup,
  TextAreaInput,
  SearchInput,
  TextInput,
  SEARCH_INPUT_ADD_ENTITY_ID,
} from "components/common/Inputs";

const PROFESSOR_SEARCH_CHANGE = "PROFESSOR_SEARCH_CHANGE";
const COURSE_SEARCH_CHANGE = "COURSE_SEARCH_CHANGE";
const SELECT_PROFESSOR = "SELECT_PROFESSOR";
const SELECT_ADD_PROFESSOR = "SELECT_ADD_PROFESSOR";
const SELECT_SEARCH_COURSES = "SELECT_SEARCH_COURSES";
const SELECT_COURSE = "SELECT_COURSE";
const SELECT_ADD_COURSE = "SELECT_ADD_COURSE";
const SELECT_NEW_COURSE = "SELECT_NEW_COURSE";
const SET_COURSE_OPTIONS = "SET_COURSE_OPTIONS";
const SET_DEPARTMENT_OPTIONS = "SET_DEPARTMENT_OPTIONS";

function reviewFormReducer(state, action) {
  switch (action.type) {
    case PROFESSOR_SEARCH_CHANGE:
      return {
        ...state,
        professorSelected: false,
      };
    case COURSE_SEARCH_CHANGE:
      return {
        ...state,
        addCourse: false,
      };
    case SELECT_PROFESSOR:
      return {
        ...state,
        professorSelected: true,
        courseOptions: [],
        addProfessor: false,
        addCourse: false,
        searchAllCourses: false,
      };
    case SET_COURSE_OPTIONS:
      return {
        ...state,
        courseOptions: action.payload,
      };
    case SET_DEPARTMENT_OPTIONS:
      return {
        ...state,
        departmentOptions: action.payload,
      };
    case SELECT_ADD_PROFESSOR:
      return {
        ...state,
        professorSelected: false,
        courseOptions: [],
        addProfessor: true,
      };
    case SELECT_SEARCH_COURSES:
      return {
        ...state,
        searchAllCourses: true,
      };
    case SELECT_COURSE:
      return {
        ...state,
        searchAllCourses: false,
        addCourse: false,
      };
    case SELECT_ADD_COURSE:
      return {
        ...state,
        addCourse: true,
      };
    case SELECT_NEW_COURSE:
      return {
        ...state,
        addCourses: false,
      };
    default:
      throw new Error(`invalid type ${action.type}`);
  }
}

export default function CreateReviewPage() {
  const history = useHistory();

  const [
    {
      professorSelected,
      courseOptions,
      addProfessor,
      addCourse,
      searchAllCourses,
      departmentOptions,
    },
    dispatch,
  ] = useReducer(reviewFormReducer, {
    professorSelected: false,
    courseOptions: [],
    addProfessor: false,
    addCourse: false,
    searchAllCourses: false,
    departmentOptions: [],
  });

  /* * * * * * * * * * * * * * * * *
   * Form contents                 *
   * * * * * * * * * * * * * * * * */

  const evaluationTexts = [
    "One of the worst experiences at Columbia. Avoid at all costs",
    "Strong negative experience. Take only if necessary",
    "Both negatives and positives much like life itself",
    "Strong positive experience. Take if possible",
    "A Columbia gem. Life-changing moments await",
  ];

  const evaluationLabels = evaluationTexts.map((label, idx) => {
    return {
      label,
      key: idx + 1,
    };
  });

  const confirmContent = (
    <>
      Your review <u>will not be published if it is</u>
      <List bulleted>
        <List.Item>Discriminatory or inappropriate</List.Item>
        <List.Item>About a nonexisting professor or course</List.Item>
        <List.Item>Libelous or defamatory</List.Item>
      </List>
      CULPA maintains the right to reject or remove any review.
    </>
  );

  /* * * * * * * * * * * * * * * * *
   * Form methods                  *
   * * * * * * * * * * * * * * * * */
  const onSubmitReview = async (data) => {
    const response = await fetch("/api/review/submit", {
      method: "POST",
      body: JSON.stringify(data),
      headers: { "Content-Type": "application/json" },
    });
    try {
      return await response.json();
    } catch (err) {
      return { error: err };
    }
  };

  const onSubmitReviewSuccess = (res) => {
    history.push(`/review/${res.reviewId}`);
  };

  /* * * * * * * * * * * * * * * * *
   * Professor Search methods      *
   * * * * * * * * * * * * * * * * */

  const searchAllCoursesId = -1;

  const onProfessorResultSelect = async ({ id: professorId }) => {
    if (professorId === SEARCH_INPUT_ADD_ENTITY_ID) {
      dispatch({ type: SELECT_ADD_PROFESSOR });
      return null;
    }

    dispatch({ type: SELECT_PROFESSOR });
    const response = await fetch(`/api/professor/${professorId}/courses`, {
      method: "GET",
      headers: { "Content-Type": "Application/json" },
    });

    try {
      const { courses } = await response.json();

      if (response.ok) {
        courses.push({
          text: `Course not listed. Add course.`,
          value: searchAllCoursesId,
          key: `Course not listed. Add course.`,
        });
        dispatch({ type: SET_COURSE_OPTIONS, payload: courses });
      }
    } catch (err) {
      return { error: err };
    }
    return null;
  };

  const onProfessorSearchChange = () => {
    dispatch({ type: PROFESSOR_SEARCH_CHANGE });
  };

  /* * * * * * * * * * * * * * * * *
   * Course Dropdown methods       *
   * * * * * * * * * * * * * * * * */

  const onOptionSelect = (courseId) => {
    courseId === searchAllCoursesId
      ? dispatch({ type: SELECT_SEARCH_COURSES })
      : dispatch({ type: SELECT_COURSE });
  };

  /* * * * * * * * * * * * * * * * *
   * Add Professor/Course Methods  *
   * * * * * * * * * * * * * * * * */

  // Prevent initial render from fetching data
  const firstRender = useRef(true);

  // Fetch departments if addProfessor or addCourse is set and not the
  // initial render.
  useEffect(() => {
    if (firstRender.current) {
      firstRender.current = false;
      return;
    }
    if (departmentOptions.length) return;
    const fetchDepartmentOptions = async () => {
      const response = await fetch(`/api/department/all?option=1`, {
        method: "GET",
        headers: { "Content-Type": "Application/json" },
      });

      const { departments } = await response.json();

      if (response.ok) {
        dispatch({ type: SET_DEPARTMENT_OPTIONS, payload: departments });
      }
    };

    fetchDepartmentOptions();
  }, [addProfessor, addCourse, departmentOptions]);

  const onCourseResultSelect = ({ id: courseId }) => {
    courseId === SEARCH_INPUT_ADD_ENTITY_ID
      ? dispatch({ type: SELECT_ADD_COURSE })
      : dispatch({ type: SELECT_NEW_COURSE });
  };

  const onCourseSearchChange = () => {
    dispatch({ type: COURSE_SEARCH_CHANGE });
  };

  /* * * * * * * * * * * * * * * * *
   * Add Professor/Course Content  *
   * * * * * * * * * * * * * * * * */

  const searchAllCoursesInput = (
    <>
      <Header>Search All Courses</Header>
      <SearchInput
        addNewEntity
        label="Select course"
        name="newCourse.search"
        rules={{ required: "Please search for a course" }}
        searchEntity="course"
        searchLimit={7}
        width={12}
        onResultSelect={onCourseResultSelect}
        onSearchChange={onCourseSearchChange}
      />
    </>
  );

  const addProfessorInputs = (
    <>
      <Header>Add New Professor</Header>
      <FormGroup>
        <TextInput
          label="New Professor First Name"
          name="newProfessor.first_name"
          rules={{ required: "Missing first name for new professor" }}
          width={6}
        />
        <TextInput
          label="New Professor Last Name"
          name="newProfessor.last_name"
          rules={{ required: "Missing last name for new professor" }}
          width={6}
        />
      </FormGroup>
      <TextInput
        label="New Professor UNI"
        name="newProfessor.uni"
        rules={{ required: "Missing professor uni" }}
        width={12}
      />
      <DropdownInput
        label="New Professor Department"
        name="newProfessor.department"
        options={departmentOptions}
        rules={{ required: "Missing department for new professor" }}
        width={12}
      />
      <SearchInput
        addNewEntity
        label="Select a course to review for the new professor"
        name="newProfessor.course"
        rules={{ required: "Missing course for new professor" }}
        searchEntity="course"
        searchLimit={7}
        width={12}
        onResultSelect={onCourseResultSelect}
        onSearchChange={onCourseSearchChange}
      />
    </>
  );

  const addCourseInputs = (
    <>
      <Header>Add New Course</Header>
      <TextInput
        label="New Course Name"
        name="newCourse.name"
        rules={{ required: "Missing name for new course" }}
        width={12}
      />
      <TextInput
        label="New Course Code"
        name="newCourse.code"
        rules={{ required: "Missing course code for new course" }}
        width={12}
      />
      <DropdownInput
        label="Select department"
        name="newCourse.department"
        options={departmentOptions}
        rules={{ required: "Please select a department" }}
        width={12}
      />
    </>
  );

  return (
    <>
      <h1>Write a Review</h1>
      <Form
        mode="onChange"
        onSubmit={onSubmitReview}
        onSuccess={onSubmitReviewSuccess}
      >
        <SearchInput
          addNewEntity
          disabled={addProfessor}
          label="Professor"
          name="professor"
          placeholder="Search for professors"
          rules={{ required: "Please select a professor" }}
          searchEntity="professor"
          searchLimit={7}
          width={6}
          onResultSelect={onProfessorResultSelect}
          onSearchChange={onProfessorSearchChange}
        />
        {professorSelected && (
          <DropdownInput
            label="Course"
            name="course"
            options={courseOptions}
            placeholder={
              professorSelected ? "" : "Please select a Professor first"
            }
            rules={{ required: "Please select a matching course" }}
            width={6}
            onOptionSelect={onOptionSelect}
          />
        )}
        {(addProfessor || addCourse || searchAllCourses) && (
          <Segment padded>
            <Grid columns={2}>
              <Grid.Column>
                {addProfessor
                  ? addProfessorInputs
                  : searchAllCourses && searchAllCoursesInput}
              </Grid.Column>
              <Grid.Column>{addCourse && addCourseInputs}</Grid.Column>
            </Grid>
          </Segment>
        )}
        <Divider section />
        <TextAreaInput
          label="Content"
          name="content"
          rows={8}
          rules={{ required: "Please describe the professor and course" }}
        />
        <TextAreaInput
          label="Workload"
          name="workload"
          rows={8}
          rules={{ required: "Please describe the workload" }}
        />
        <RadioInputGroup labels={evaluationLabels} name="evaluation" />
        <SubmitConfirm
          content={confirmContent}
          header="Are you sure you want to submit this review?"
        />
      </Form>
    </>
  );
}
