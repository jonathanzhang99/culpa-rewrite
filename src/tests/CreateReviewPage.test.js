import { render, fireEvent, act, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import CreateReviewPage from "components/CreateReviewPage";

describe("Create Review Page tests", () => {
  const serverProfessorResponseJson = {
    // TODO (Sungbin): please change this to match
    searchResults: [
      {
        title: "Nakul Verma",
        description: "ml",
        content: "testContent1",
      },
      {
        title: "Lee Bollinger",
        description: "law",
        content: "testContent2",
      },
    ],
  };

  const serverCoursesResponseJson = {
    courses: [
      { text: "Machine Learning", value: 888, key: "Machine Learning" },
      {
        text: "Computational Learning Theory",
        value: 889,
        key: "Computational Learning Theory",
      },
    ],
  };

  const serverSubmissionResponseJson = { reviewId: 99 };

  const serverProfessorResponse = () =>
    Promise.resolve({
      ok: true,
      json: () => serverProfessorResponseJson,
    });

  const serverCoursesResponse = () =>
    Promise.resolve({
      ok: true,
      json: () => serverCoursesResponseJson,
    });

  const serverSubmissionResponse = () =>
    Promise.resolve({
      ok: true,
      json: () => serverSubmissionResponseJson,
    });

  test("should render", () => {
    const snapshot = render(<CreateReviewPage />);
    expect(snapshot).toMatchSnapshot();
  });

  describe("frontend interactions", () => {
    const fieldErrorLabels = [
      "Please select a professor",
      "Please select a matching course",
      "Please describe the professor and course",
      "Please describe the workload",
    ];

    beforeEach(() => {
      render(
        <MemoryRouter>
          <CreateReviewPage />
        </MemoryRouter>
      );
    });

    test("empty fields display error and lack of modal", async () => {
      await act(async () => {
        fireEvent.click(screen.getByRole("button"));
      });
      fieldErrorLabels.forEach(async (errorLabel) => {
        expect(await screen.getByText(errorLabel)).toBeInTheDocument();
      });
      expect(
        await screen.queryByText(/Are you sure you want to submit this review/i)
      ).not.toBeInTheDocument();
    });

    describe("select a professor", () => {
      let mockFetch;

      beforeEach(() => {
        mockFetch = jest.spyOn(global, "fetch");
        mockFetch.mockImplementation(serverProfessorResponse);
      });

      afterEach(() => jest.resetAllMocks());

      test("professor query too short", async () => {
        await act(async () => {
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "a" },
          });
        });
        expect(
          await screen.getByText(/no professors found/i)
        ).toBeInTheDocument();
        expect(mockFetch).not.toHaveBeenCalled();
      });

      test("valid professor query", async () => {
        await act(async () => {
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "testProfessorName" },
          });
        });

        expect(await screen.getByText(/nakul verma/i)).toBeInTheDocument();
        expect(await screen.getByText(/lee bollinger/i)).toBeInTheDocument();
        expect(mockFetch).toHaveBeenCalledTimes(1);
        expect(mockFetch).toHaveBeenCalledWith(
          "/api/search?searchEntity=professors&query=testProfessorName",
          {
            method: "GET",
            headers: { "Content-Type": "Application/json" },
          }
        );
      });

      test("no professors when query returns no data", async () => {
        mockFetch.mockImplementation(() =>
          Promise.resolve({
            ok: true,
            json: () => {
              return { searchResults: [] };
            },
          })
        );

        await act(async () => {
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "testProfessorName" },
          });
        });
        expect(
          await screen.getByText(/no professors found/i)
        ).toBeInTheDocument();
        expect(mockFetch).toHaveBeenCalledTimes(1);
        expect(mockFetch).toHaveBeenCalledWith(
          "/api/search?searchEntity=professors&query=testProfessorName",
          {
            method: "GET",
            headers: { "Content-Type": "Application/json" },
          }
        );
      });
    });

    describe("select a course", () => {
      let mockFetch;

      beforeEach(() => {
        mockFetch = jest.spyOn(global, "fetch");
        mockFetch
          .mockImplementationOnce(serverProfessorResponse)
          .mockImplementationOnce(serverCoursesResponse);
      });

      afterEach(() => jest.clearAllMocks());

      test("courses are not displayed with no professor", async () => {
        expect(
          await screen.getByText(/please select a professor first/i)
        ).toBeInTheDocument();
        expect(mockFetch).not.toHaveBeenCalled();
      });

      test("courses are displayed on professor select", async () => {
        await act(async () =>
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "testProfessorName" },
          })
        );

        await act(async () =>
          fireEvent.click(screen.getByText(/nakul verma/i))
        );

        expect(await screen.getByText(/machine learning/i)).toBeInTheDocument();
        expect(
          await screen.getByText(/computational learning theory/i)
        ).toBeInTheDocument();
        expect(mockFetch).toHaveBeenCalledTimes(2);
        // TODO: (Sungbin, JZ) the following line will not properly run without defining the
        // backend data format for Search Results
        // expect(mockFetch).toHaveBeenLastCalledWith("/api/professor/1/courses", {
        //   method: "GET",
        //   headers: { "Content-Type": "Application/json" },
        // });
      });
    });

    describe("on valid form inputs", () => {
      let mockFetch;

      beforeEach(async () => {
        mockFetch = jest.spyOn(global, "fetch");
        mockFetch
          .mockImplementationOnce(serverProfessorResponse)
          .mockImplementationOnce(serverCoursesResponse)
          .mockImplementationOnce(serverSubmissionResponse);

        /**
         * Fill out form with fake data
         *
         * professor: "nakul verma",
         * course: "Machine Learning",
         * content: "this class was great!!!",
         * workload: "suffered so much!!!",
         * evaluation: 3
         *
         */
        await act(async () =>
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "testProfessorName" },
          })
        );

        await act(async () =>
          fireEvent.click(screen.getByText(/nakul verma/i))
        );

        await act(async () => {
          fireEvent.click(screen.getByText(/machine learning/i));
          fireEvent.input(screen.getByRole("textbox", { name: /content/i }), {
            target: { value: "this class was great!!!" },
          });
          fireEvent.input(screen.getByRole("textbox", { name: /workload/i }), {
            target: { value: "suffered so much!!!" },
          });
          fireEvent.click(
            screen.getByLabelText(
              /Both negatives and positives much like life itself/i
            )
          );
        });
      });

      afterEach(() => jest.resetAllMocks());

      test("no errors are shown", async () => {
        await act(async () => {
          fireEvent.click(screen.getByRole("button"));
        });

        fieldErrorLabels.forEach(async (errorLabel) => {
          expect(await screen.queryByText(errorLabel)).not.toBeInTheDocument();
        });
        // expect submission to not have been called
        expect(mockFetch).toHaveBeenCalledTimes(2);
      });

      test("modal appears", async () => {
        await act(async () => {
          fireEvent.click(screen.getByRole("button"));
        });

        expect(
          await screen.queryByText(
            /Are you sure you want to submit this review/i
          )
        ).toBeInTheDocument();
      });

      test("properly submits data", async () => {
        await act(async () => fireEvent.click(screen.getByRole("button")));
        await act(async () =>
          fireEvent.click(screen.getAllByRole("button", { name: /submit/i })[1])
        );
        expect(mockFetch).toHaveBeenCalledTimes(3);
        // TODO (Sungbin): Change to match backend
        expect(mockFetch).toHaveBeenLastCalledWith("/api/review/submit", {
          method: "POST",
          body: JSON.stringify({
            professor: {
              title: "Nakul Verma",
              description: "ml",
              content: "testContent1",
            },
            course: 888,
            content: "this class was great!!!",
            workload: "suffered so much!!!",
            evaluation: 3,
          }),
          headers: { "Content-Type": "application/json" },
        });
      });

      test("modal can be closed", async () => {
        await act(async () => fireEvent.click(screen.getByRole("button")));
        await act(async () =>
          fireEvent.click(screen.getByRole("button", { name: /edit review/i }))
        );
        expect(mockFetch).toHaveBeenCalledTimes(2);
        expect(
          await screen.queryByText(
            /Are you sure you want to submit this review/i
          )
        ).not.toBeInTheDocument();
      });

      test("fields can be revalidated after initial validation", async () => {
        // first open and close the modal
        await act(async () => fireEvent.click(screen.getByRole("button")));
        await act(async () =>
          fireEvent.click(screen.getByRole("button", { name: /edit review/i }))
        );

        // reset all fields
        await act(async () =>
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "" },
          })
        );

        await act(async () => {
          fireEvent.input(screen.getByRole("textbox", { name: /content/i }), {
            target: { value: "" },
          });
          fireEvent.input(screen.getByRole("textbox", { name: /workload/i }), {
            target: { value: "" },
          });
          await act(async () => fireEvent.click(screen.getByRole("button")));
        });

        fieldErrorLabels.forEach(async (errorLabel) => {
          // TODO BUG (JZ): Currently, the form does not support resetting the course
          if (errorLabel !== "Please select a matching course") {
            expect(await screen.getByText(errorLabel)).toBeInTheDocument();
          }
        });
        expect(mockFetch).toHaveBeenCalledTimes(2);
      });
    });

    /**
     * This test ensures that the user can edit their review and change
     * any of the valid fields.
     */
    describe("change inputs after initial validation", () => {
      let mockFetch;

      beforeEach(async () => {
        mockFetch = jest.spyOn(global, "fetch");
        mockFetch
          .mockImplementationOnce(serverProfessorResponse)
          .mockImplementationOnce(serverCoursesResponse)
          .mockImplementationOnce(serverProfessorResponse)
          .mockImplementationOnce(serverCoursesResponse)
          .mockImplementationOnce(serverSubmissionResponse);

        await act(async () =>
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "testProfessorName" },
          })
        );

        await act(async () =>
          fireEvent.click(screen.getByText(/nakul verma/i))
        );

        await act(async () => {
          fireEvent.click(screen.getByText(/machine learning/i));
          fireEvent.input(screen.getByRole("textbox", { name: /content/i }), {
            target: { value: "this class was great!!!" },
          });
          fireEvent.input(screen.getByRole("textbox", { name: /workload/i }), {
            target: { value: "suffered so much!!!" },
          });
          fireEvent.click(
            screen.getByLabelText(
              /Both negatives and positives much like life itself/i
            )
          );
        });
      });

      test("fields can changed after initial validation", async () => {
        // first open and close the modal
        await act(async () => fireEvent.click(screen.getByRole("button")));
        await act(async () =>
          fireEvent.click(screen.getByRole("button", { name: /edit review/i }))
        );

        // change all fields to new values
        await act(async () => {
          fireEvent.input(screen.getByRole("textbox", { name: /professor/i }), {
            target: { value: "anotherTestProfessor" },
          });
        });

        await act(async () =>
          fireEvent.click(screen.getByText(/lee bollinger/i))
        );

        await act(async () => {
          fireEvent.click(screen.getByText(/computational learning theory/i));
          fireEvent.input(screen.getByRole("textbox", { name: /content/i }), {
            target: { value: "why tf is bollinger teaching clt" },
          });
          fireEvent.input(screen.getByRole("textbox", { name: /workload/i }), {
            target: {
              value: "too many case studies not enough chernoff-hoeffding",
            },
          });
          fireEvent.click(
            screen.getByLabelText(
              "One of the worst experiences at Columbia. Avoid at all costs"
            )
          );
        });
        await act(async () => fireEvent.click(screen.getByRole("button")));

        // no errors should be present
        fieldErrorLabels.forEach(async (errorLabel) => {
          expect(await screen.queryByText(errorLabel)).not.toBeInTheDocument();
        });

        // submit the modal confirmation
        await act(async () =>
          fireEvent.click(screen.getAllByRole("button", { name: /submit/i })[1])
        );
        expect(mockFetch).toHaveBeenCalledTimes(5);
        // TODO (Sungbin): Change to match backend
        expect(mockFetch).toHaveBeenLastCalledWith("/api/review/submit", {
          method: "POST",
          body: JSON.stringify({
            professor: {
              title: "Lee Bollinger",
              description: "law",
              content: "testContent2",
            },
            course: 889,
            content: "why tf is bollinger teaching clt",
            workload: "too many case studies not enough chernoff-hoeffding",
            evaluation: 1,
          }),
          headers: { "Content-Type": "application/json" },
        });
      });
    });
  });
});
