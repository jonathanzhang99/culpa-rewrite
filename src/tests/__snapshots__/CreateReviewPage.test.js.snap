// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Review Page tests should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Write a Review
      </h1>
      <form
        class="ui form"
      >
        <div
          class="ui error message"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              Error
            </div>
          </div>
        </div>
        <div
          class="six wide field"
        >
          <label
            for="form-input-professor"
          >
            Professor
          </label>
          <div
            class="ui fluid search"
          >
            <div
              class="ui icon input"
            >
              <input
                aria-invalid="true"
                autocomplete="off"
                class="prompt"
                id="form-input-professor"
                name="professor"
                tabindex="0"
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                class="search icon"
              />
            </div>
            <div
              class="results transition"
            >
              <div
                class="message empty"
              >
                <div
                  class="header"
                >
                  No professors found
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ui section divider"
        />
        <div
          class="field"
        >
          <label
            for="form-input-content"
          >
            Content
          </label>
          <textarea
            aria-invalid="true"
            aria-label="Content"
            id="form-input-content"
            name="content"
            rows="8"
          />
        </div>
        <div
          class="field"
        >
          <label
            for="form-input-workload"
          >
            Workload
          </label>
          <textarea
            aria-invalid="true"
            aria-label="Workload"
            id="form-input-workload"
            name="workload"
            rows="8"
          />
        </div>
        <div
          class="inline unstackable equal width fields"
        >
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
            >
              <input
                class="hidden"
                id="evaluation-radio-button-1"
                name="evaluation"
                readonly=""
                tabindex="0"
                type="radio"
                value="1"
              />
              <label
                for="evaluation-radio-button-1"
              >
                One of the worst experiences at Columbia. Avoid at all costs
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
            >
              <input
                class="hidden"
                id="evaluation-radio-button-2"
                name="evaluation"
                readonly=""
                tabindex="0"
                type="radio"
                value="2"
              />
              <label
                for="evaluation-radio-button-2"
              >
                Strong negative experience. Take only if necessary
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
            >
              <input
                class="hidden"
                id="evaluation-radio-button-3"
                name="evaluation"
                readonly=""
                tabindex="0"
                type="radio"
                value="3"
              />
              <label
                for="evaluation-radio-button-3"
              >
                Both negatives and positives much like life itself
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
            >
              <input
                class="hidden"
                id="evaluation-radio-button-4"
                name="evaluation"
                readonly=""
                tabindex="0"
                type="radio"
                value="4"
              />
              <label
                for="evaluation-radio-button-4"
              >
                Strong positive experience. Take if possible
              </label>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="ui radio checkbox"
            >
              <input
                class="hidden"
                id="evaluation-radio-button-5"
                name="evaluation"
                readonly=""
                tabindex="0"
                type="radio"
                value="5"
              />
              <label
                for="evaluation-radio-button-5"
              >
                A Columbia gem. Life-changing moments await
              </label>
            </div>
          </div>
        </div>
        <button
          class="ui button"
          name="confirmModal"
          type="button"
        >
          SUBMIT
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <h1>
      Write a Review
    </h1>
    <form
      class="ui form"
    >
      <div
        class="ui error message"
      >
        <div
          class="content"
        >
          <div
            class="header"
          >
            Error
          </div>
        </div>
      </div>
      <div
        class="six wide field"
      >
        <label
          for="form-input-professor"
        >
          Professor
        </label>
        <div
          class="ui fluid search"
        >
          <div
            class="ui icon input"
          >
            <input
              aria-invalid="true"
              autocomplete="off"
              class="prompt"
              id="form-input-professor"
              name="professor"
              tabindex="0"
              type="text"
              value=""
            />
            <i
              aria-hidden="true"
              class="search icon"
            />
          </div>
          <div
            class="results transition"
          >
            <div
              class="message empty"
            >
              <div
                class="header"
              >
                No professors found
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ui section divider"
      />
      <div
        class="field"
      >
        <label
          for="form-input-content"
        >
          Content
        </label>
        <textarea
          aria-invalid="true"
          aria-label="Content"
          id="form-input-content"
          name="content"
          rows="8"
        />
      </div>
      <div
        class="field"
      >
        <label
          for="form-input-workload"
        >
          Workload
        </label>
        <textarea
          aria-invalid="true"
          aria-label="Workload"
          id="form-input-workload"
          name="workload"
          rows="8"
        />
      </div>
      <div
        class="inline unstackable equal width fields"
      >
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              id="evaluation-radio-button-1"
              name="evaluation"
              readonly=""
              tabindex="0"
              type="radio"
              value="1"
            />
            <label
              for="evaluation-radio-button-1"
            >
              One of the worst experiences at Columbia. Avoid at all costs
            </label>
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              id="evaluation-radio-button-2"
              name="evaluation"
              readonly=""
              tabindex="0"
              type="radio"
              value="2"
            />
            <label
              for="evaluation-radio-button-2"
            >
              Strong negative experience. Take only if necessary
            </label>
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              id="evaluation-radio-button-3"
              name="evaluation"
              readonly=""
              tabindex="0"
              type="radio"
              value="3"
            />
            <label
              for="evaluation-radio-button-3"
            >
              Both negatives and positives much like life itself
            </label>
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              id="evaluation-radio-button-4"
              name="evaluation"
              readonly=""
              tabindex="0"
              type="radio"
              value="4"
            />
            <label
              for="evaluation-radio-button-4"
            >
              Strong positive experience. Take if possible
            </label>
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="ui radio checkbox"
          >
            <input
              class="hidden"
              id="evaluation-radio-button-5"
              name="evaluation"
              readonly=""
              tabindex="0"
              type="radio"
              value="5"
            />
            <label
              for="evaluation-radio-button-5"
            >
              A Columbia gem. Life-changing moments await
            </label>
          </div>
        </div>
      </div>
      <button
        class="ui button"
        name="confirmModal"
        type="button"
      >
        SUBMIT
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
